tensorboard/loader.py

```diff
@@ -732,85 +735,305 @@ def get_next_event(self):
+            (offset, log.rowid, offset))
 
   def mark(self):
     """Marks current position in file so reset() can be called."""
     if self._prepended_events:
       raise ValueError('mark() offsets must be monotonic')
-    self._mark = self._offset
+    self._mark = self._get_offset()
     self._saved_events.clear()
 
   def reset(self):
     """Resets read state to where mark() was called."""
     if self._mark == -1:
-      raise ValueError('mark() was not called')
+      return
     self._prepended_events.extend(self._saved_events)
     self._saved_events.clear()
 
   def close(self):
-    """Closes event log reader if open.
+    """Closes all event log readers.
 
-    Further i/o is not permitted after this method is called.
+    This method may be called multiple times, but further operations
+    are not permitted.
+
+    Raises:
+      Exception: To propagate the most recent exception thrown by the
+          EventLogReader close method. Suppressed exceptions are
+          logged.
     """
-    if self._reader is not None:
-      self._reader.close()
-      self._reader = None
+    util.close_all(self._logs)
+    self._index = len(self._logs)
+    self._mark = -1
+    self._prepended_events.clear()
+    self._saved_events.clear()
+
+  def _cleanup(self):
+    # Last event log has to be preserved so we can continue enforcing
+    # monotonicity. We entomb offset because that also has to be
+    # monotonic, but the size does not.
+    if 0 < self._index < len(self._logs):
+      deleted = self._logs[:self._index]
+      self._logs = self._logs[self._index:]
+      self._index = 0
+      self._entombed_progress += sum(l.get_offset() for l in deleted)
+      util.close_all(deleted)
+
+  def _skip_to_event_log(self, i):
+    should_mark = self._mark != -1 and i > self._index
+    self._index = i
+    if should_mark:
+      self._prepended_events.clear()
+      self.mark()
 
   def __hash__(self):
-    return hash(self._key)
+    return hash(self.rowid)
 
   def __eq__(self, other):
-    return self._key == other._key
+    return self.rowid == other.rowid
 
   def __lt__(self, other):
-    return self._key < other._key
+    return self.rowid < other.rowid
 
   def __str__(self):
     offset = self.get_offset()
     if offset:
-      return u'EventLog{path=%s, offset=%d}' % (self.path, offset)
+      return u'RunReader{name=%s, offset=%d}' % (self.name, offset)
     else:
-      return u'EventLog{%s}' % self.path
+      return u'RunReader{%s}' % self.name
+
+
+def _get_basename(path):
+  """Gets base name of path.
+
+  This is the same as os.path.basename, however it may potentially do
+  i/o to handle a few edge cases, which would otherwise cause the
+  result to be less meaningful, e.g. "." and "..".
+
+  :type path: str
+  :rtype: str
+  """
+  result = os.path.basename(os.path.normpath(path))
+  if result in ('', '.', '..'):
+    result = os.path.basename(os.path.realpath(path))
```

Comment:
Good idea. It seems this whole thing can be written `os.path.basename(os.path.normpath(os.path.join(os.getcwd(), path)))`.